giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+     * @param edges Number of edges loaded.
+     * @param vertices Number of vertices loaded.
+     * @param verticesProcessed Number of vertices processed.
+     * @param bytesReceived Number of bytes received.
+     * @param oocBytesInjected Number of bytes stored/loaded due to OOC.
+     */
+    public void addRecord(long memUsed, long edges, long vertices,
+                          long verticesProcessed,
+                          long bytesReceived, long oocBytesInjected) {
+      checkState(memUsed > 0, "Memory Usage cannot be negative");
+      if (dataSamples.size() > 0) {
+        double[] last = dataSamples.get(dataSamples.size() - 1);
+        if (edges == last[0] && vertices == last[1] &&
+          verticesProcessed == last[2] && bytesReceived == last[3] &&
+          oocBytesInjected == last[4]) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug(
+              "addRecord: avoiding to add the same entry as the last one!");
+          }
+          return;
+        }
+      }
+      dataSamples.add(new double[] {edges, vertices, verticesProcessed,
+        bytesReceived, oocBytesInjected});
+      memorySamples.add((double) memUsed);
+
+      // Weed out the columns that are all zero
+      validColumnIndices.clear();
+      for (int i = 0; i < 5; ++i) {
+        boolean validIndex = false;
+        // Check if there is a non-zero entry in the column
+        for (double[] value : dataSamples) {
+          if (value[i] != 0) {
+            validIndex = true;
+            break;
+          }
+        }
+        if (validIndex) {
+          // check if all entries are not equal to each other
+          double firstValue = -1;
+          boolean allEqual = true;
+          for (double[] value : dataSamples) {
+            if (firstValue == -1) {
+              firstValue = value[i];
+            } else {
+              if (Math.abs((value[i] - firstValue) / firstValue) > 0.01) {
+                allEqual = false;
+                break;
+              }
+            }
+          }
+          validIndex = !allEqual;
+          if (validIndex) {
+            // Check if the column has linear dependency with another column
+            for (int col = i + 1; col < 5; ++col) {
+              if (isLinearDependence(dataSamples, i, col)) {
+                validIndex = false;
+                break;
+              }
+            }
+          }
+        }
+
+        if (validIndex) {
+          validColumnIndices.add(i);
+        }
+      }
+
+      // If we filtered out columns in the previous step, we are going to run
+      // the regression without those columns.
+
+      // Create the coefficient table
+      boolean setIsValid = false;
+      lock.lock();
+      try {
+        if (validColumnIndices.size() >= 1 &&
+          dataSamples.size() >= validColumnIndices.size() + 1) {
+
+          double[][] xValues = new double[dataSamples.size()][];
+          double[] yValues = new double[memorySamples.size()];
```

Comment:
I maybe miss something, but how do you guarantee that another thread will not add more values to memorySamples after you've created xValues but before you created yValues?  In other words, how do you guarantee that xValues and yValues have the same size?